File: adapters/aduro.py
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "homeostat",
#     "paho-mqtt>=2,<3",
# ]
#
# [tool.uv.sources]
# homeostat = { path = "../sdk/python", editable = true }
# ///
"""Aduro pellet-burner adapter (docs/design.md, "Burners and interlocks
(settled 2026-09-09, #37, #38)").

The burner speaks the NBE UDP protocol; github.com/freol35241/aduro2mqtt
bridges it to MQTT. The bridge polls on a fixed interval
(ADURO_POLL_INTERVAL, default 30 s) and republishes EVERY topic each cycle,
changed or not: {base}/status (the `status *` response, positional CSV
mapped onto pyduro's STATUS_PARAMS names — 116 fields — as one JSON
object, values floated where they parse), {base}/operating and
{base}/advanced (one JSON object each), {base}/settings/{group} (17 groups
of device configuration), {base}/consumption/{key} (9 arrays) and
{base}/logs. Commands go to {base}/set as {"path": "<group>.<name>",
"value": ...}. The entity file's `id` is the bridge's MQTT_BASE_TOPIC
(`aduro2mqtt` by default); the file stem is the entity name. One entity
per burner.

What becomes state, and how much. Republish-on-poll means an adapter
forwarding each message would record every field 2,700 times a day
whether or not it ever moved — the status topic alone is two and a half
times the reporting house's entire heat-pump adapter (#37). So the
adapter subscribes exactly {base}/status and {base}/operating, and
publishes a field ONLY WHEN ITS VALUE CHANGES (plus once, on the first
poll after start; late joiners read the core's state mirror). Settings
are configuration, not samples; consumption, advanced and logs are the
long tail — none are subscribed. Status fields publish under their
firmware names (dots included: `regulation.fixed_power` is a legal key
segment); operating fields publish as `operating_{field}`, the prefix
keeping the two NBE namespaces apart without a table of one to check
the other against.

Four normalizations carry the `burner` vocabulary:

- `on` (the base aspect, bool) is DERIVED from the run state: the burner
  is on unless `state` is one of OFF_STATES. Start and stop are momentary
  writes in this dialect (misc.start / misc.stop), so `on` can only ever
  be the device's own readback, never an echo of a command. OFF_STATES
  holds the single code observed so far — 14, the burner idle and unlit
  through the survey that settled #37 — and grows as the heating season
  produces codes; `state` and `substate` pass through raw beside it for
  exactly that purpose. A code not in OFF_STATES reads as on.
- `power_level` (feature) is status `regulation.fixed_power`, the fixed
  output setting, 10 / 50 / 100 on this device — published as an int
  when integral so the enum matches.
- `flue_temperature` is status `smoke_temp` (the exhaust reading; the
  flue cutouts at the reporting house read this one).
- `boiler_temperature` is status `boiler_temp`.

Everything else passes through, including `shaft_temp` (the feed shaft,
the device's own fire-safety reading) and `power_pct` (actual output) —
except a raw field that would mint a reserved name (`available`, or one
of the normalized names above arriving under its bus name rather than
its firmware name), which drops with a "reserved-aspect" event, and a
field name that is not a legal key segment, which drops with
"malformed-payload"; the rest of the document still publishes.

Commands (COMMANDS) — the burner is an arbitrated entity (the family's
`on` and an automation's `power_level` lease independently, per aspect),
so every command arrives on home/arbiter/{room}/{entity}/{aspect}:

- `on`: strictly a bool. true -> {"path": "misc.start", "value": "1"},
  false -> {"path": "misc.stop", "value": "1"}, the bridge's own switch
  shape.
- `power_level`: strictly the integer 10, 50 or 100 ->
  {"path": "regulation.fixed_power", "value": <int>}.

Anything else — a wrong type, an out-of-enum level, an unknown aspect,
a malformed or envelope-less payload — DROPS with an "invalid-command"
(or "malformed-payload") health event and never reaches the device.

Discovery is the static one-record-per-entity document (the ivt490
shape): base-topic id, a suggested `burner` capability stanza with the
`power_level` feature, the aspect descriptor (ASPECT_FIELDS) and a
`bound` flag that flips true the first time the base topic is seen.

Availability: the bridge publishes on a cadence, so silence is the loss
signal — a receive timer flips home/state/{room}/{entity}/available to
false after availability_timeout_s (parameter, owner-editable, default
300 s, about nine polls) without a message from the base topic, with one
"device-silent" health event per down transition; the next message flips
it back. The bridge skips a topic's publish when the burner does not
answer, so an unreachable burner and a dead bridge both go silent. On
loss every other aspect stands — stale, never false.

Operational note: the bridge's README wires Home Assistant switches and
selects straight to {base}/set; disable them (and any Node-RED writer)
before this adapter goes live — one master per device.
"""

import json
import os
import threading
import time

import homeostat
from homeostat import house, keys, mqtt
from homeostat.params import LiveParams

PARAM_DEFAULTS = {"availability_timeout_s": 300.0}
_UNSET = object()


class Params(LiveParams):
    """availability_timeout_s from home/config/{unit}/*, live."""

    @property
    def availability_timeout_s(self) -> float:
        return max(0.1, self.get("availability_timeout_s"))


# Status field -> burner vocabulary (see module docstring).
ASPECT_OVERRIDES = {
    "smoke_temp": "flue_temperature",
    "boiler_temp": "boiler_temperature",
    "regulation.fixed_power": "power_level",
}

# Run-state codes that mean the burner is not making heat. 14 is the one
# code observed to date (idle, unlit); extend from the heating season.
OFF_STATES = frozenset({14})

# Names a status field may not mint raw: the adapter's own liveness
# signal, and the vocabulary the overrides derive (a wire field literally
# named `on` would otherwise overwrite the derived one and poison the
# publish-on-change cache).
RESERVED_STATUS_FIELDS = frozenset({"available", "on", *ASPECT_OVERRIDES.values()})

POWER_LEVELS = (10, 50, 100)

# Commandable aspect -> the NBE set path (None for `on`, whose path
# depends on the value: misc.start / misc.stop).
COMMANDS = {"on": None, "power_level": "regulation.fixed_power"}

# The aspect descriptor (docs/design.md, Aspect descriptors). The dashboard
# renders descriptor commands as enums or numbers, so `on` is described as
# a two-valued enum — a segmented off/on control on the card. Labels keep
# the firmware field name in parentheses where it differs.
ASPECT_GROUPS = ["control", "readings", "status"]
T = "temperature"
ASPECT_FIELDS = {
    "on": {
        "label": "burner",
        "kind": "boolean",
        "group": "control",
        "values": [{"value": False, "label": "off"}, {"value": True, "label": "on"}],
        "command": {"type": "enum", "editable_by": "family"},
    },
    "power_level": {
        "label": "power",
        "kind": "enum",
        "group": "control",
        "values": [{"value": level, "label": f"{level}%"} for level in POWER_LEVELS],
        "command": {"type": "enum", "editable_by": "family"},
    },
    "flue_temperature": {"label": "flue (smoke_temp)", "kind": T, "group": "readings"},
    "boiler_temperature": {"label": "boiler (boiler_temp)", "kind": T, "group": "readings"},
    "shaft_temp": {"label": "feed shaft (shaft_temp)", "kind": T, "group": "readings"},
    "power_pct": {"label": "output (power_pct)", "kind": "percent", "group": "readings"},
    "state": {"label": "run state (state)", "kind": "number", "group": "status"},
    "substate": {"label": "run substate (substate)", "kind": "number", "group": "status"},
}
ASPECT_DESCRIPTOR = {"schema": 1, "groups": ASPECT_GROUPS, "fields": ASPECT_FIELDS}


def status_aspects(status: dict) -> tuple[dict, list[str]]:
    """One status document to the bus aspects it yields: the normalized
    names, the derived `on`, and every other field under its firmware
    name — plus the raw fields dropped for naming a reserved aspect.
    `power_level` is an int when the wire float is integral."""
    aspects = {}
    reserved = []
    for field, value in status.items():
        if field in RESERVED_STATUS_FIELDS:
            reserved.append(field)
            continue
        aspect = ASPECT_OVERRIDES.get(field, field)
        if aspect == "power_level" and isinstance(value, float) and value.is_integer():
            value = int(value)
        aspects[aspect] = value
    state = status.get("state")
    if isinstance(state, (int, float)):  # a run-state code is a number
        aspects["on"] = state not in OFF_STATES
    return aspects, reserved


def command_body(aspect: str, value):
    """The {base}/set payload for a validated command, or None when the
    value is not one this aspect takes (a bool for `on`, one of
    POWER_LEVELS — an int, never a bool — for `power_level`)."""
    if aspect == "on":
        if not isinstance(value, bool):
            return None
        return {"path": "misc.start" if value else "misc.stop", "value": "1"}
    if aspect == "power_level":
        if isinstance(value, bool) or not isinstance(value, int) or value not in POWER_LEVELS:
            return None
        return {"path": COMMANDS[aspect], "value": value}
    return None


def route(topic: str, entities):
    """The bound entity and the topic's segments past its base-topic
    prefix, or (None, None)."""
    for entity in entities:
        prefix = f"{entity.id}/"
        if topic.startswith(prefix):
            return entity, topic[len(prefix) :].split("/")
    return None, None


def main():
    unit = os.environ[keys.ENV_UNIT]
    config = house.load_adapter(unit)
    endpoint = mqtt.parse_endpoint(config.endpoint)

    session = homeostat.connect()
    params = Params(session, PARAM_DEFAULTS)
    seen: set[str] = set()

    # Publish-on-change: the last value put on the bus per (entity, aspect).
    last: dict[tuple[str, str], object] = {}

    availability_lock = threading.Lock()
    last_rx = {e.id: time.monotonic() for e in config.entities}
    available: dict[str, bool] = {}

    def set_available(entity, value: bool) -> bool:
        with availability_lock:
            if available.get(entity.name) == value:
                return False
            available[entity.name] = value
            session.put_json(keys.state_key(entity.room, entity.name, "available"), value)
        return True

    def inventory():
        return [
            {
                "id": e.id,
                "configured": True,
                "entity": e.name,
                "bound": e.id in seen,
                "suggested": {"capability": "burner", "features": ["power_level"]},
                "aspects": ASPECT_DESCRIPTOR,
            }
            for e in config.entities
        ]

    def on_aduro_message(client, userdata, msg):
        entity, rest = route(msg.topic, config.entities)
        if entity is None or len(rest) != 1 or rest[0] not in ("status", "operating"):
            return

        last_rx[entity.id] = time.monotonic()
        set_available(entity, True)
        if entity.id not in seen:
            seen.add(entity.id)
            session.put_json(keys.discovery_key(unit), inventory())

        try:
            document = json.loads(msg.payload)
            if not isinstance(document, dict):
                raise ValueError("not an object")
        except ValueError:
            session.health_event("drop", reason="malformed-payload", topic=msg.topic)
            return

        if rest[0] == "status":
            aspects, reserved = status_aspects(document)
            for field in reserved:
                session.health_event("drop", reason="reserved-aspect", topic=msg.topic, field=field)
        else:
            aspects = {f"operating_{field}": value for field, value in document.items()}
        for aspect, value in aspects.items():
            try:
                key = keys.state_key(entity.room, entity.name, aspect)
            except ValueError:
                session.health_event(
                    "drop", reason="malformed-payload", topic=msg.topic, field=aspect
                )
                continue
            if last.get((entity.name, aspect), _UNSET) == value:
                continue  # unchanged since the last poll: not a sample
            last[(entity.name, aspect)] = value
            session.put_json(key, value)

    def cmd_handler(entity):
        def handler(sample):
            parsed = session.parse_command(sample)
            if parsed is None:
                return
            aspect, value = parsed
            body = command_body(aspect, value) if aspect in COMMANDS else None
            if body is None:
                session.health_event(
                    "drop", reason="invalid-command", key=str(sample.key_expr), aspect=aspect, value=value
                )
                return
            client.publish(f"{entity.id}/set", json.dumps(body))

        return handler

    topics = [
        (topic, 0)
        for e in config.entities
        for topic in (f"{e.id}/status", f"{e.id}/operating")
    ]
    client = mqtt.connect(endpoint, on_aduro_message, topics)

    subscribers = [
        session.subscribe(expr, cmd_handler(e))
        for e in config.entities
        for expr in keys.command_keyexprs(e)
    ]

    session.put_json(keys.discovery_key(unit), inventory())

    stop = threading.Event()

    def watchdog():
        while True:
            timeout = params.availability_timeout_s
            if stop.wait(min(1.0, timeout / 4)):
                return
            now = time.monotonic()
            for entity in config.entities:
                if now - last_rx[entity.id] > timeout and set_available(entity, False):
                    session.health_event("device-silent", topic=entity.id)

    watchdog_thread = threading.Thread(target=watchdog, daemon=True)
    watchdog_thread.start()

    session.ready()

    mqtt.wait_for_shutdown()

    stop.set()
    watchdog_thread.join(timeout=5)
    # The MQTT loop stops first: an in-flight on_message during teardown
    # would otherwise put on a closed zenoh session.
    client.loop_stop()
    client.disconnect()
    for sub in subscribers:
        sub.undeclare()
    session.close()


if __name__ == "__main__":
    main()

File: tests/aduro.rs
//! Aduro pellet-burner adapter integration tests: each scenario spawns a
//! real mosquitto broker on a free port plus the real supervisor on the
//! aduro fixture house, and asserts on both buses.

mod common;

use std::time::Duration;

use homeostat::bus::HealthStatus;
use serde_json::json;
use zenoh::sample::SampleKind;

use common::{
    assert_unit_contract, await_health, expect_drop_event, expect_event_kind, expect_states,
    health_watch, Mosquitto, Mqtt, Supervisor,
};

const FIXTURE: &str = "tests/fixture_house_aduro";
const PORT_ENV: &str = "HOMEOSTAT_TEST_MQTT_PORT";
const BASE: &str = "aduro2mqtt"; // the fixture entity's base topic (its `id`)
const EVENT_KEY: &str = "home/health/aduro/event";
const ON_CMD_KEY: &str = "home/cmd/livingroom/burner/on";
const ON_ARBITER_KEY: &str = "home/arbiter/livingroom/burner/on";
const POWER_CMD_KEY: &str = "home/cmd/livingroom/burner/power_level";
const SET_TOPIC: &str = "aduro2mqtt/set";
const AVAILABLE_KEY: &str = "home/state/livingroom/burner/available";

/// A status document the bridge would publish: the burner idle (state 14)
/// at 10% fixed power, with the readings the vocabulary normalizes and a
/// few passthrough fields, floated as the bridge floats them.
fn status(smoke_temp: f64) -> String {
    json!({
        "boiler_temp": 21.4,
        "smoke_temp": smoke_temp,
        "shaft_temp": 23.1,
        "state": 14.0,
        "substate": 0.0,
        "power_pct": 0.0,
        "regulation.fixed_power": 10.0,
        "city": "Somewhere",
    })
    .to_string()
}

async fn setup() -> (Mosquitto, Supervisor, zenoh::Session) {
    let mosquitto = Mosquitto::spawn();
    let port = mosquitto.port.to_string();
    let sup = Supervisor::spawn_with_env(FIXTURE, &[(PORT_ENV, &port)]);
    let observer = sup.observer().await;
    let token_sub = observer
        .liveliness()
        .declare_subscriber("home/health/aduro/alive")
        .history(true)
        .await
        .expect("liveliness subscriber");
    let token = tokio::time::timeout(Duration::from_secs(60), token_sub.recv_async())
        .await
        .expect("adapter liveliness token within 60s")
        .expect("liveliness stream open");
    assert_eq!(token.kind(), SampleKind::Put);
    (mosquitto, sup, observer)
}

/// Waits for the arbiter to be running: commands to the arbitrated burner
/// only reach the adapter through it.
async fn await_arbiter(observer: &zenoh::Session) {
    let mut watch = health_watch(observer, "arbiter").await;
    await_health(&mut watch, Duration::from_secs(60), |h| {
        h.status == HealthStatus::Running
    })
    .await;
}

/// (a) One status document fans out to the burner vocabulary (`on` derived
/// from the run state, `power_level` as an int, the two temperatures
/// normalized) plus passthrough under firmware names, dotted or not; an
/// operating document publishes prefixed; the identical republish a poll
/// later yields NOTHING, a changed field yields exactly that field; a
/// non-object payload drops with malformed-payload; and the discovery
/// record carries the descriptor.
#[tokio::test(flavor = "multi_thread")]
async fn status_translates_on_change_only() {
    let (mosquitto, mut sup, observer) = setup().await;
    let state_sub = observer
        .declare_subscriber("home/state/livingroom/burner/*")
        .await
        .expect("state subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-state").await;

    mqtt.publish(&format!("{BASE}/status"), &status(25.5)).await;
    mqtt.publish(&format!("{BASE}/operating"), r#"{"boiler_temp": 21.4}"#)
        .await;
    // Unsubscribed topics: nothing, not even an event.
    mqtt.publish(
        &format!("{BASE}/settings/regulation"),
        r#"{"fixed_power": 10.0}"#,
    )
    .await;
    mqtt.publish(&format!("{BASE}/consumption/counter"), "[1.0, 2.0]")
        .await;

    expect_states(
        &state_sub,
        &[
            ("home/state/livingroom/burner/available", json!(true)),
            ("home/state/livingroom/burner/on", json!(false)),
            ("home/state/livingroom/burner/power_level", json!(10)),
            ("home/state/livingroom/burner/flue_temperature", json!(25.5)),
            (
                "home/state/livingroom/burner/boiler_temperature",
                json!(21.4),
            ),
            ("home/state/livingroom/burner/shaft_temp", json!(23.1)),
            ("home/state/livingroom/burner/state", json!(14.0)),
            ("home/state/livingroom/burner/city", json!("Somewhere")),
            (
                "home/state/livingroom/burner/operating_boiler_temp",
                json!(21.4),
            ),
        ],
    )
    .await;

    // Drain: the first poll may still be delivering; then republish the
    // same document and require silence.
    while tokio::time::timeout(Duration::from_millis(500), state_sub.recv_async())
        .await
        .is_ok()
    {}
    mqtt.publish(&format!("{BASE}/status"), &status(25.5)).await;
    let silence = tokio::time::timeout(Duration::from_millis(1500), state_sub.recv_async()).await;
    assert!(
        silence.is_err(),
        "unchanged status republished to the bus: {silence:?}"
    );

    // One field moves: exactly one sample.
    mqtt.publish(&format!("{BASE}/status"), &status(26.0)).await;
    let sample = tokio::time::timeout(Duration::from_secs(10), state_sub.recv_async())
        .await
        .expect("changed field within 10s")
        .expect("state stream open");
    assert_eq!(
        sample.key_expr().as_str(),
        "home/state/livingroom/burner/flue_temperature"
    );
    let silence = tokio::time::timeout(Duration::from_millis(1500), state_sub.recv_async()).await;
    assert!(
        silence.is_err(),
        "unchanged fields republished: {silence:?}"
    );

    let no_event = tokio::time::timeout(Duration::from_millis(500), event_sub.recv_async()).await;
    assert!(no_event.is_err(), "unexpected health event: {no_event:?}");

    // A non-object payload is malformed.
    mqtt.publish(&format!("{BASE}/status"), "[1, 2, 3]").await;
    expect_drop_event(&event_sub, "malformed-payload").await;

    let replies = observer
        .get("home/discovery/aduro")
        .await
        .expect("discovery query");
    let mut inventory = None;
    while let Ok(reply) = replies.recv_async().await {
        if let Ok(sample) = reply.result() {
            inventory =
                serde_json::from_slice::<serde_json::Value>(&sample.payload().to_bytes()).ok();
        }
    }
    let inventory = inventory.expect("discovery inventory served");
    let record = &inventory[0];
    assert_eq!(record["entity"], json!("burner"), "{inventory}");
    assert_eq!(record["id"], json!(BASE));
    assert_eq!(record["bound"], json!(true));
    assert_eq!(
        record["suggested"],
        json!({"capability": "burner", "features": ["power_level"]})
    );
    let fields = &record["aspects"]["fields"];
    assert_eq!(
        record["aspects"]["groups"],
        json!(["control", "readings", "status"])
    );
    assert_eq!(
        fields["on"]["command"],
        json!({"type": "enum", "editable_by": "family"})
    );
    assert_eq!(
        fields["on"]["values"][1],
        json!({"value": true, "label": "on"})
    );
    assert_eq!(
        fields["power_level"]["values"][1],
        json!({"value": 50, "label": "50%"})
    );
    assert_eq!(
        fields["flue_temperature"]["label"],
        json!("flue (smoke_temp)")
    );
    assert!(
        fields.get("city").is_none(),
        "undescribed aspects are simply absent"
    );

    sup.shutdown();
}

/// (a2) Receive-timer availability: a status message flips available =
/// true, availability_timeout_s (5 s in the fixture) of silence flips it
/// false with a "device-silent" health event, the next message flips it
/// back.
#[tokio::test(flavor = "multi_thread")]
async fn silence_flips_available() {
    let (mosquitto, mut sup, observer) = setup().await;
    let state_sub = observer
        .declare_subscriber(AVAILABLE_KEY)
        .await
        .expect("state subscriber");
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-availability").await;

    mqtt.publish(&format!("{BASE}/status"), &status(25.5)).await;
    expect_states(&state_sub, &[(AVAILABLE_KEY, json!(true))]).await;

    expect_states(&state_sub, &[(AVAILABLE_KEY, json!(false))]).await;
    expect_event_kind(&event_sub, "device-silent").await;

    mqtt.publish(&format!("{BASE}/status"), &status(25.5)).await;
    expect_states(&state_sub, &[(AVAILABLE_KEY, json!(true))]).await;

    sup.shutdown();
}

/// (b) Manual-band `on` and `power_level` wishes ride the arbiter to
/// {base}/set in the bridge's own command shape; while the manual `on`
/// lease holds, an automation-band `on` is refused upstream — but the
/// automation's `power_level` still passes, because leases are per aspect.
#[tokio::test(flavor = "multi_thread")]
async fn commands_reach_set_topic_via_arbiter_per_aspect() {
    let (mosquitto, mut sup, observer) = setup().await;
    await_arbiter(&observer).await;
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-cmd").await;
    mqtt.subscribe(SET_TOPIC).await;

    let wish = json!({"value": true, "priority": "manual", "actor": "test"});
    observer
        .put(ON_CMD_KEY, wish.to_string())
        .await
        .expect("cmd put");
    let (topic, payload) = mqtt
        .next_message(Duration::from_secs(10))
        .await
        .expect("start publish");
    assert_eq!(topic, SET_TOPIC);
    assert_eq!(
        serde_json::from_slice::<serde_json::Value>(&payload).unwrap(),
        json!({"path": "misc.start", "value": "1"})
    );

    let wish = json!({"value": false, "priority": "manual", "actor": "test"});
    observer
        .put(ON_CMD_KEY, wish.to_string())
        .await
        .expect("cmd put");
    let (_, payload) = mqtt
        .next_message(Duration::from_secs(10))
        .await
        .expect("stop publish");
    assert_eq!(
        serde_json::from_slice::<serde_json::Value>(&payload).unwrap(),
        json!({"path": "misc.stop", "value": "1"})
    );

    // The manual `on` lease holds (hold_minutes = 30): an automation `on`
    // is refused upstream and never reaches MQTT.
    let wish = json!({"value": true, "priority": "automation", "actor": "heat_plan"});
    observer
        .put(ON_CMD_KEY, wish.to_string())
        .await
        .expect("cmd put");
    let silence = mqtt.next_message(Duration::from_millis(1500)).await;
    assert!(
        silence.is_none(),
        "refused automation wish reached MQTT: {silence:?}"
    );

    // Per-aspect leases: the same automation's power_level passes.
    let wish = json!({"value": 50, "priority": "automation", "actor": "heat_plan"});
    observer
        .put(POWER_CMD_KEY, wish.to_string())
        .await
        .expect("cmd put");
    let (_, payload) = mqtt
        .next_message(Duration::from_secs(10))
        .await
        .expect("power publish");
    assert_eq!(
        serde_json::from_slice::<serde_json::Value>(&payload).unwrap(),
        json!({"path": "regulation.fixed_power", "value": 50})
    );

    sup.shutdown();
}

/// (c) Validation: a non-bool `on`, an out-of-enum or non-integer
/// `power_level`, an unknown aspect and an envelope-less payload all drop
/// with the right event and nothing reaches {base}/set.
#[tokio::test(flavor = "multi_thread")]
async fn invalid_commands_drop_with_events() {
    let (mosquitto, mut sup, observer) = setup().await;
    await_arbiter(&observer).await;
    let event_sub = observer
        .declare_subscriber(EVENT_KEY)
        .await
        .expect("event subscriber");
    let mut mqtt = Mqtt::connect(mosquitto.port, "test-invalid").await;
    mqtt.subscribe(SET_TOPIC).await;

    let wish = json!({"value": 1, "priority": "manual", "actor": "test"});
    observer
        .put(ON_CMD_KEY, wish.to_string())
        .await
        .expect("cmd put");
    let event = expect_drop_event(&event_sub, "invalid-command").await;
    assert_eq!(event["aspect"], json!("on"));
    assert_eq!(event["value"], json!(1));

    for value in [json!(30), json!(50.5), json!("50"), json!(true)] {
        let wish = json!({"value": value, "priority": "manual", "actor": "test"});
        observer
            .put(POWER_CMD_KEY, wish.to_string())
            .await
            .expect("cmd put");
        let event = expect_drop_event(&event_sub, "invalid-command").await;
        assert_eq!(event["aspect"], json!("power_level"));
        assert_eq!(event["value"], value);
    }

    let wish = json!({"value": 200.0, "priority": "manual", "actor": "test"});
    observer
        .put("home/cmd/livingroom/burner/smoke_temp", wish.to_string())
        .await
        .expect("cmd put");
    let event = expect_drop_event(&event_sub, "invalid-command").await;
    assert_eq!(event["aspect"], json!("smoke_temp"));

    observer.put(ON_ARBITER_KEY, "true").await.expect("cmd put");
    expect_drop_event(&event_sub, "invalid-command").await;

    let silence = mqtt.next_message(Duration::from_millis(1500)).await;
    assert!(
        silence.is_none(),
        "an invalid command reached MQTT: {silence:?}"
    );

    sup.shutdown();
}

/// (e) The adapter honors the unit contract: liveliness token when ready,
/// clean SIGTERM shutdown within the grace, no orphans.
#[tokio::test(flavor = "multi_thread")]
async fn adapter_honors_unit_contract() {
    let (_mosquitto, mut sup, observer) = setup().await;
    assert_unit_contract(&mut sup, &observer, "aduro").await;
}
